parquet-common/src/main/java/org/apache/parquet/SemanticVersion.java

```diff
@@ -141,7 +186,91 @@ public int hashCode() {
 
   @Override
   public String toString() {
-    return major + "." + minor + "." + patch;
+    StringBuilder sb = new StringBuilder();
+    sb.append(major).append(".").append(minor).append(".").append(patch);
+    if (prerelease) {
+      sb.append(unknown);
+    }
+    if (pre != null) {
+      sb.append(pre.original);
+    }
+    if (buildInfo != null) {
+      sb.append(buildInfo);
+    }
+    return sb.toString();
+  }
+
+  private static class NumberOrString implements Comparable<NumberOrString> {
+    private static final Pattern NUMERIC = Pattern.compile("\\d+");
+
+    private final String original;
+    private final boolean isNumeric;
+    private final int number;
+
+    public NumberOrString(String numberOrString) {
+      this.original = numberOrString;
+      this.isNumeric = NUMERIC.matcher(numberOrString).matches();
+      if (isNumeric) {
+        this.number = Integer.parseInt(numberOrString);
+      } else {
+        this.number = -1;
+      }
+    }
+
+    @Override
+    public int compareTo(NumberOrString that) {
+      // Numeric identifiers always have lower precedence than non-numeric identifiers.
+      int cmp = compareBooleans(that.isNumeric, this.isNumeric);
+      if (cmp != 0) {
+        return cmp;
+      }
+
+      if (isNumeric) {
+        // identifiers consisting of only digits are compared numerically
+        return this.number - that.number;
```

Comment:
can you use Integer.compare() here, or just inline it as written here:
https://github.com/openjdk-mirror/jdk7u-jdk/blob/master/src/share/classes/java/lang/Integer.java#L1021

the x - y trick has issues w/ overflow